### DISCUSSION TITLE
Help caching specific files in Cefsharp.

### DISCUSSION THREAD

**Discussion by quinguyenhuu** [cefsharp/CefSharp#3489] [General]
Is there any way to cache an array of requests in Cefsharp? I want to cache all the files that have big size such as svg, js, video... and doesn't want to cache all the files that have small size. Please help.

**Comment by amaitland** [comment]
>  Is there any way to cache an array of requests in Cefsharp?

Not specifically no.

You can use the standard http cache headers.

**Reply by quinguyenhuu** [reply]
Thanks. I want to calculate the size of each request, and if the request size is too big, i plan to save the response to the local and then next time if the same request, I just load the response from the local instead of sending request over internet. 
Are there any functions that i can override to help me solve this problem? 
Please help. Thanks.

**Reply by amaitland** [reply]
That should be possible. 

1. First check if the file exists on disk and return a resource handler if true.

https://github.com/cefsharp/CefSharp/wiki/General-Usage#request-interception

You can use https://cefsharp.github.io/api/89.0.x/html/M_CefSharp_ResourceHandler_FromFilePath.htm to load from disk.

2. Use a IResponseFilter to copy the response into a stream. 
https://github.com/cefsharp/CefSharp/wiki/General-Usage#response-filtering

The example shows converting the response to a string, you can write the stream to disk.

In  GetResourceResponseFilter you'll hopefully have access to a response headers and be able to obtain the Content-Length to determine if the size meets your minimum requirement. Return null in GetResourceResponseFilter if you don't want to get the response body. Content-length header is optional from memory, so no guarantees that a response will have it.

Streaming video will be a problem as you won't have an entire response to load from disk.

If you need to support range requests when using a ResourceHandler then you'll need a custom implementation. I have a basic working example at https://github.com/cefsharp/CefSharp/blob/cefsharp/89/CefSharp.Example/AcceptRangeResourceHandler.cs

**Reply by quinguyenhuu** [reply]
Very thank you.

**Reply by amaitland** [reply]
You'll probably want to implement `2.` first, the order I listed is just based on order of events the methods are called at runtime.

**Reply by quinguyenhuu** [reply]
yes, thank you so much.
